src/es5.js

```diff
@@ -144,11 +158,166 @@ function proxyProperty(draft, prop, enumerable) {
+	size: state => latest(state).size,
+	has: state => key => latest(state).has(key),
+	set: state => (key, value) => {
+		if (latest(state).get(key) !== value) {
+			prepareCopy(state)
+			markChanged(state)
+			state.assigned.set(key, true)
+			state.copy.set(key, value)
+		}
+		return state.draft
+	},
+	delete: state => key => {
+		prepareCopy(state)
+		markChanged(state)
+		state.assigned.set(key, false)
+		state.copy.delete(key)
+		return false
+	},
+	clear: state => () => {
+		if (!state.copy) {
+			prepareCopy(state)
+		}
+		markChanged(state)
+		state.assigned = new Map()
+		for (const key of latest(state).keys()) {
+			state.assigned.set(key, false)
+		}
+		return state.copy.clear()
+	},
+	forEach: (state, key, reciever) => cb => {
+		latest(state).forEach((value, key, map) => {
+			cb(reciever.get(key), key, map)
+		})
+	},
+	get: state => key => {
+		const value = latest(state).get(key)
+
+		if (state.finalizing || state.finalized || !isDraftable(value)) {
+			return value
+		}
+
+		if (value !== state.base.get(key)) {
+			return value
+		}
+		const draft = createProxy(value, state)
+		prepareCopy(state)
+		state.copy.set(key, draft)
+		return draft
+	},
+	keys: state => () => latest(state).keys(),
+	values: iterateMapValues,
+	entries: iterateMapValues
+})
+
+function proxySet(target) {
+	Object.defineProperties(target, setTraps)
+
+	if (hasSymbol) {
+		Object.defineProperty(
+			target,
+			Symbol.iterator,
+			proxyMethod(iterateSetValues)
+		)
+	}
+}
+
+const iterateSetValues = makeIterateSetValues(createProxy)
+
+const setTraps = finalizeTraps({
+	size: state => {
+		return latest(state).size
+	},
+	add: state => value => {
+		if (!latest(state).has(value)) {
+			markChanged(state)
+			if (!state.copy) {
+				prepareCopy(state)
+			}
+			state.copy.add(value)
+		}
+		return state.draft
+	},
+	delete: state => value => {
+		markChanged(state)
+		if (!state.copy) {
+			prepareCopy(state)
+		}
+		return state.copy.delete(value)
+	},
+	has: state => key => {
+		return latest(state).has(key)
+	},
+	clear: state => () => {
+		markChanged(state)
+		if (!state.copy) {
+			prepareCopy(state)
+		}
+		return state.copy.clear()
+	},
+	keys: iterateSetValues,
+	entries: iterateSetValues,
+	values: iterateSetValues,
+	forEach: state => (cb, thisArg) => {
+		const iterator = iterateSetValues(state)()
+		let result = iterator.next()
+		while (!result.done) {
+			cb.call(thisArg, result.value, result.value, state.draft)
+			result = iterator.next()
+		}
+	}
+})
+
+function finalizeTraps(traps) {
+	return Object.keys(traps).reduce(function(acc, key) {
+		const builder = key === "size" ? proxyAttr : proxyMethod
+		acc[key] = builder(traps[key], key)
+		return acc
+	}, {})
+}
+
+function proxyAttr(fn) {
+	return {
+		get() {
+			const state = this[DRAFT_STATE]
+			assertUnrevoked(state)
+			return fn(state)
+		}
+	}
+}
+
+function proxyMethod(trap, key) {
+	return {
+		get() {
+			return (...args) => {
+				const state = this[DRAFT_STATE]
```

Comment:
Using `this` in an arrow function, is this doing what it is supposed to be doing? Should be `function (...args)` I presume

Reply:
If this is correct, a comment would be warranted :)

Reply:
@mweststrate it provides context for `this`. When we call `someMap.forEach()`, it call `get() {...}` which return function

Reply:
Hmm, I've tried to use function, and it works. I have no idea why)

Reply:
Updated, just in case. Don't get why it worked though :-P.